druid/src/menu/mod.rs

```diff
@@ -0,0 +1,762 @@
+            (NeedsRebuild, _) | (_, NeedsRebuild) => NeedsRebuild,
+            (NeedsRefresh, _) | (_, NeedsRefresh) => NeedsRefresh,
+            _ => UpToDate,
+        }
+    }
+}
+
+/// This is the trait that enables recursive visiting of all menu entries. It isn't publically
+/// visible (the publically visible analogue of this is `Into<MenuEntry<T>>`).
+trait MenuTrait<T> {
+    /// Called when a menu item is activated.
+    ///
+    /// `id` is the id of the entry that got activated. If this is different from your id, you are
+    /// responsible for routing the activation to your child items.
+    fn activate(&mut self, ctx: &mut MenuEventCtx, id: MenuItemId, data: &mut T, env: &Env);
+
+    /// Called when the data is changed.
+    fn update(&mut self, old_data: &T, data: &T, env: &Env) -> MenuUpdate;
+
+    /// Called to refresh the menu.
+    fn refresh(&mut self, ctx: &mut MenuBuildCtx, data: &T, env: &Env);
+}
+
+/// A wrapper for a menu item (or submenu) to give it access to a part of its parent data.
+///
+/// This is the menu analogue of [`LensWrap`]. You will usually create it with [`Menu::lens`] or
+/// [`MenuItem::lens`] instead of using this struct directly.
+///
+/// [`LensWrap`]: crate::widget::LensWrap
+pub struct MenuLensWrap<L, U> {
+    lens: L,
+    inner: Box<dyn MenuTrait<U>>,
+}
+
+impl<T: Data, U: Data, L: Lens<T, U>> MenuTrait<T> for MenuLensWrap<L, U> {
+    fn activate(&mut self, ctx: &mut MenuEventCtx, id: MenuItemId, data: &mut T, env: &Env) {
+        let inner = &mut self.inner;
+        self.lens
+            .with_mut(data, |u| inner.activate(ctx, id, u, env));
+    }
+
+    fn update(&mut self, old_data: &T, data: &T, env: &Env) -> MenuUpdate {
+        let inner = &mut self.inner;
+        let lens = &self.lens;
+        lens.with(old_data, |old| {
+            lens.with(data, |new| inner.update(old, new, env))
+        })
+    }
+
+    fn refresh(&mut self, ctx: &mut MenuBuildCtx, data: &T, env: &Env) {
+        let inner = &mut self.inner;
+        self.lens.with(data, |u| inner.refresh(ctx, u, env))
+    }
+}
+
+impl<T: Data, U: Data, L: Lens<T, U> + 'static> From<MenuLensWrap<L, U>> for MenuEntry<T> {
+    fn from(m: MenuLensWrap<L, U>) -> MenuEntry<T> {
+        MenuEntry { inner: Box::new(m) }
+    }
+}
+
+/// An entry in a menu.
+///
+/// An entry is either a [`MenuItem`], a submenu (i.e. [`Menu`]), or one of a few other
+/// possibilities (such as one of the two options above, wrapped in a [`MenuLensWrap`]).
+pub struct MenuEntry<T> {
+    inner: Box<dyn MenuTrait<T>>,
+}
+
+type MenuPredicate<T> = Box<dyn FnMut(&T, &T, &Env) -> bool>;
+
+/// A menu.
+///
+/// Menus can be nested arbitrarily, so this could also be a submenu.
+/// See the [module level documentation](crate::menu) for more on how to use menus.
+pub struct Menu<T> {
+    needs_rebuild: Option<MenuPredicate<T>>,
+    needs_refresh: Option<MenuPredicate<T>>,
+    item: MenuItem<T>,
+    children: Vec<MenuEntry<T>>,
+    // bloom?
```

Comment:
This would require children to have some sort of identity, and for there to be some benefit in only visiting specific children. Do you think this would make sense?

Reply:
The `MenuItem`s are supposed to have unique ids, so blooms would basically allow us to traverse directly to the right `MenuItem` when getting a menu command from the platform.

I didn't do it yet because:
- the menu probably can't be big enough (or commands common enough) for traversing the menu tree to be a bottleneck, and
- it might actually be more expensive to construct bloom filters each time the menu is rebuilt (I have more to say about rebuilding, but I'll put that elsewhere later)